File: rust/crates/openmelon-tui/src/project.rs
use std::fs;
use std::path::{Path, PathBuf};

use anyhow::{bail, Context, Result};
use serde::{Deserialize, Serialize};

pub const STATE_DIR: &str = ".openmelon";
pub const PROJECT_FILE: &str = "project.json";
pub const OUTPUTS_DIR: &str = "outputs";

#[derive(Debug, Clone, Deserialize, Serialize, Default)]
pub struct Project {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub persona: String,
    #[serde(default)]
    pub constraints: Vec<String>,
    #[serde(default)]
    pub defaults: Defaults,
    #[serde(default)]
    pub providers: std::collections::BTreeMap<String, ProviderConfig>,
    #[serde(default)]
    pub settings: Settings,
}

#[derive(Debug, Clone, Deserialize, Serialize, Default)]
pub struct Defaults {
    #[serde(default)]
    pub llm_provider: String,
    #[serde(default)]
    pub llm_model: String,
    #[serde(default)]
    pub image_provider: String,
    #[serde(default)]
    pub image_model: String,
    #[serde(default)]
    pub vision_model: String,
    #[serde(default)]
    pub locale: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Default)]
pub struct ProviderConfig {
    #[serde(default)]
    pub api_key: String,
    #[serde(default)]
    pub base_url: String,
}

#[derive(Debug, Clone, Deserialize, Serialize, Default)]
pub struct Settings {
    #[serde(default)]
    pub bash_permission_mode: String,
    #[serde(default)]
    pub reasoning_effort: String,
}

#[derive(Debug, Clone)]
pub struct Workspace {
    pub root: PathBuf,
    pub project: Project,
}

impl Workspace {
    pub fn discover(start: impl AsRef<Path>) -> Result<Self> {
        let root = discover_root(start)?;
        let project = load_project(&root)?;
        ensure_project_dirs(&root)?;
        Ok(Self { root, project })
    }

    pub fn state_dir(&self) -> PathBuf {
        self.root.join(STATE_DIR)
    }

    pub fn outputs_dir(&self) -> PathBuf {
        self.root.join(OUTPUTS_DIR)
    }

    pub fn session_outputs_dir(&self, session_id: &str) -> PathBuf {
        self.outputs_dir().join("sessions").join(session_id)
    }

    pub fn artifact_outputs_dir(&self, slug: &str, timestamp: &str) -> PathBuf {
        self.outputs_dir()
            .join("artifacts")
            .join(slug)
            .join(timestamp)
    }
}

pub fn discover_root(start: impl AsRef<Path>) -> Result<PathBuf> {
    let mut cur = start
        .as_ref()
        .canonicalize()
        .with_context(|| format!("resolve {}", start.as_ref().display()))?;

    loop {
        if cur.join(STATE_DIR).join(PROJECT_FILE).is_file() {
            return Ok(cur);
        }
        if !cur.pop() {
            bail!(
                "not an openmelon project: no {}/{} found from {} upward",
                STATE_DIR,
                PROJECT_FILE,
                start.as_ref().display()
            );
        }
    }
}

pub fn load_project(root: &Path) -> Result<Project> {
    let path = root.join(STATE_DIR).join(PROJECT_FILE);
    let body = fs::read_to_string(&path).with_context(|| format!("read {}", path.display()))?;
    serde_json::from_str(&body).with_context(|| format!("parse {}", path.display()))
}

pub fn update_project_json(root: &Path, update: impl FnOnce(&mut serde_json::Value)) -> Result<()> {
    let path = root.join(STATE_DIR).join(PROJECT_FILE);
    let body = fs::read_to_string(&path).with_context(|| format!("read {}", path.display()))?;
    let mut value: serde_json::Value =
        serde_json::from_str(&body).with_context(|| format!("parse {}", path.display()))?;
    update(&mut value);
    fs::write(&path, serde_json::to_string_pretty(&value)? + "\n")
        .with_context(|| format!("write {}", path.display()))?;
    Ok(())
}

pub fn set_project_default(root: &Path, key: &str, value: &str) -> Result<()> {
    update_project_json(root, |project| {
        ensure_object(project, "defaults");
        project["defaults"][key] = serde_json::Value::String(value.to_string());
    })
}

pub fn set_project_setting(root: &Path, key: &str, value: &str) -> Result<()> {
    update_project_json(root, |project| {
        ensure_object(project, "settings");
        if value.trim().is_empty() {
            if let Some(obj) = project["settings"].as_object_mut() {
                obj.remove(key);
            }
        } else {
            project["settings"][key] = serde_json::Value::String(value.to_string());
        }
    })
}

fn ensure_object(value: &mut serde_json::Value, key: &str) {
    if !value.get(key).is_some_and(serde_json::Value::is_object) {
        value[key] = serde_json::json!({});
    }
}

pub fn ensure_project_dirs(root: &Path) -> Result<()> {
    for sub in [
        "characters",
        "references",
        "materials",
        "sessions",
        "spaces",
    ] {
        fs::create_dir_all(root.join(STATE_DIR).join(sub))
            .with_context(|| format!("create .openmelon/{sub}"))?;
    }
    fs::create_dir_all(root.join(OUTPUTS_DIR)).context("create outputs dir")?;
    Ok(())
}

pub fn resolve_output_dir(root: &Path, requested: &str, fallback: &Path) -> Result<PathBuf> {
    let candidate = if requested.trim().is_empty() {
        fallback.to_path_buf()
    } else {
        let requested = PathBuf::from(requested.trim());
        if requested.is_absolute() {
            requested
        } else {
            root.join(requested)
        }
    };

    let root = root.canonicalize()?;
    let state = root.join(STATE_DIR);
    let abs = if candidate.exists() {
        candidate.canonicalize()?
    } else {
        let parent = candidate.parent().unwrap_or(&root);
        let parent_abs = if parent.exists() {
            parent.canonicalize()?
        } else {
            root.clone()
        };
        parent_abs.join(candidate.file_name().unwrap_or_default())
    };

    if !abs.starts_with(&root) {
        bail!("output dir {} escapes project workdir", candidate.display());
    }
    if abs.starts_with(&state) {
        bail!(
            "output dir {} is inside .openmelon; choose a visible project directory",
            candidate.display()
        );
    }
    Ok(abs)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn resolve_output_rejects_hidden_state() {
        let root = std::env::temp_dir().join("openmelon-rust-project");
        let _ = std::fs::create_dir_all(root.join(".openmelon"));
        let fallback = root.join("outputs");
        let err = resolve_output_dir(&root, ".openmelon/artifacts", &fallback).unwrap_err();
        assert!(err.to_string().contains(".openmelon"));
    }
}

File: tui/src/core/project.ts
import {promises as fs} from 'node:fs';
import path from 'node:path';
import {findUp, readJsonFile, writeJsonFile} from './fs.js';

export type ProjectDefaults = {
	llm_provider?: string;
	llm_model?: string;
	image_provider?: string;
	image_model?: string;
	vision_model?: string;
	locale?: string;
};

export type ProviderConfig = {
	api_key?: string;
	base_url?: string;
};

export type ProjectSettings = {
	bash_permission_mode?: 'strict' | 'auto' | 'trusted';
	reasoning_effort?: 'none' | 'minimal' | 'low' | 'medium' | 'high' | 'xhigh';
};

export type ProjectConfig = {
	id: string;
	name: string;
	description?: string;
	persona?: string;
	constraints?: string[];
	defaults?: ProjectDefaults;
	providers?: Record<string, ProviderConfig>;
	settings?: ProjectSettings;
	created_at?: string;
};

export const stateDirName = '.openmelon';
export const projectFileName = 'project.json';

export async function discoverProject(start = process.cwd()) {
	return findUp(start, path.join(stateDirName, projectFileName));
}

export async function loadProject(workdir: string) {
	return readJsonFile<ProjectConfig>(projectPath(workdir), {id: '', name: ''});
}

export async function saveProject(workdir: string, project: ProjectConfig) {
	await writeJsonFile(projectPath(workdir), project);
}

export function projectPath(workdir: string) {
	return path.join(workdir, stateDirName, projectFileName);
}

export function stateDir(workdir: string) {
	return path.join(workdir, stateDirName);
}

export function outputsDir(workdir: string) {
	return path.join(workdir, 'outputs');
}

/** Visible output directory for one generation session. */
export function sessionOutputDir(workdir: string, sessionId: string) {
	const id = sessionId.trim() || 'session';
	return path.join(outputsDir(workdir), 'sessions', id);
}

/** Default visible directory for a promoted artifact bucket. */
export function artifactOutputDir(workdir: string, slug: string, timestamp: string) {
	return path.join(outputsDir(workdir), 'artifacts', slug, timestamp);
}

function pathInside(parent: string, child: string) {
	const rel = path.relative(parent, child);
	return rel === '' || (!rel.startsWith('..') && !path.isAbsolute(rel));
}

/**
 * Resolve a user/model-selected output directory inside the project. Empty
 * `requested` uses `fallback`, then <workdir>/outputs. Paths under .openmelon
 * are rejected (reserved for internal state); paths escaping the workdir too.
 */
export function resolveOutputDir(workdir: string, requested: string, fallback: string) {
	const absWorkdir = path.resolve(workdir);
	const req = requested.trim();
	let out = fallback;
	if (req) {
		out = path.isAbsolute(req) ? req : path.join(absWorkdir, req);
	}
	if (!out.trim()) {
		out = outputsDir(absWorkdir);
	}
	const absOut = path.resolve(out);
	if (!pathInside(absWorkdir, absOut)) {
		throw new Error(`output dir ${JSON.stringify(requested)} escapes project workdir`);
	}
	if (pathInside(stateDir(absWorkdir), absOut)) {
		throw new Error(`output dir ${JSON.stringify(requested)} is inside .openmelon; choose a visible project directory`);
	}
	return absOut;
}

export async function initProject(workdir: string, project: ProjectConfig) {
	const existing = await discoverProject(workdir);
	if (existing === path.resolve(workdir)) {
		return {created: false};
	}
	for (const subdir of ['characters', 'references', 'materials', 'sessions', 'spaces']) {
		await fs.mkdir(path.join(stateDir(workdir), subdir), {recursive: true});
	}
	await fs.mkdir(outputsDir(workdir), {recursive: true});
	await writeJsonFile(projectPath(workdir), project);
	await ensureProjectGitignore(workdir);
	return {created: true};
}

export function slugFromBase(base: string) {
	const slug = base
		.toLowerCase()
		.replace(/[^a-z0-9._ -]/g, '')
		.replace(/[._ -]+/g, '-')
		.replace(/^-+|-+$/g, '');
	const normalized = /^[a-z]/.test(slug) ? slug : `project-${slug}`.replace(/-+$/g, '');
	return normalized.length >= 2 ? normalized : 'project';
}

export function validateProjectId(id: string) {
	if (id.length < 2 || id.length > 64) {
		throw new Error(`project id ${JSON.stringify(id)} must be 2..64 chars`);
	}
	if (!/^[a-z][a-z0-9-]*$/.test(id)) {
		throw new Error(`project id ${JSON.stringify(id)} must be kebab-case`);
	}
	if (id.endsWith('-') || id.includes('--')) {
		throw new Error(`project id ${JSON.stringify(id)} must not end with '-' or contain '--'`);
	}
}

export async function ensureProjectGitignore(workdir: string) {
	const target = path.join(stateDir(workdir), '.gitignore');
	try {
		await fs.access(target);
		return;
	} catch {}
	await fs.mkdir(stateDir(workdir), {recursive: true});
	await fs.writeFile(target, gitignoreContent);
}

const gitignoreContent = `# openmelon - auto-generated. Edit if you want different defaults.
# These paths are relative to this .openmelon/ directory.

# API keys.
credentials.json

# Per-run conversation transcripts + generated images.
sessions/

# Legacy hidden outputs. User-facing outputs go under ../outputs/.
artifacts/
`;
